lib/api/core/validation/index.js

```diff
@@ -0,0 +1,788 @@
+  Promise = require('bluebird'),
+  Dsl = require('../../dsl'),
+  defaultFieldSpecification = {
+    'mandatory': false,
+    'multivalued': {
+      value: false
+    }
+  },
+  defaultTypesFiles = [
+    'anything',
+    'boolean',
+    'date',
+    'email',
+    'enum',
+    'geoPoint',
+    'geoShape',
+    'integer',
+    'ipAddress',
+    'numeric',
+    'object',
+    'string',
+    'url'
+  ];
+
+/**
+ * @param {Kuzzle} kuzzle
+ * @constructor
+ */
+function Validation (kuzzle) {
+  /** @type {Kuzzle} */
+  this.kuzzle = kuzzle;
+
+  /** @type {...ValidationType} */
+  this.types = {};
+
+  /** @type {string[]} */
+  this.typeAllowsChildren = [];
+
+  /** @type {DocumentSpecification} */
+  this.specification = {};
+
+  /** @type {Dsl} */
+  this.dsl = new Dsl();
+
+  this.rawConfiguration = {};
+}
+
+/**
+ * Walks through all types in "defaultTypesFiles" initializes all types
+ */
+Validation.prototype.init = function () {
+  var
+    ValidationType;
+
+  defaultTypesFiles.forEach(typeFile => {
+    ValidationType = require(`./types/${typeFile}`);
+    this.addType(new ValidationType());
+  });
+};
+
+/**
+ * @param {RequestObject} requestObject
+ * returns {Promise}
+ */
+Validation.prototype.validate = function (requestObject) {
+  var
+    index = requestObject.index,
+    collection = requestObject.collection,
+    controller = requestObject.controller,
+    action = requestObject.action,
+    id;
+
+  if (!requestObject.hasOwnProperty('data')) {
+    return new BadRequestError('The request object must provide data');
+  }
+
+  if (!requestObject.data.hasOwnProperty('body')) {
+    return new BadRequestError('The request object must provide a document body');
+  }
+
+
+  if (controller === 'write' && action === 'update') {
+    if (!requestObject.data.hasOwnProperty('_id')) {
+      return new BadRequestError('Update request must provide an _id.');
+    }
+
+    id = requestObject.data._id;
+  }
+
+  return this.validationPromise(index, collection, controller, action, id, requestObject.data.body, false)
+    .then(validationObject => {
+      if (!validationObject.validation) {
+        throw new BadRequestError(validationObject.errorMessages.join('\n'));
+      }
+
+      // We only modify the requestObject if the validation succeeds
+      requestObject.data.body = validationObject.documentBody;
+
+      return requestObject;
+    });
+};
+
+/**
+ * @param {string} index
+ * @param {string} collection
+ * @param {string} controller
+ * @param {string} action
+ * @param {string} id
+ * @param {*} documentCandidate
+ * @param {string[]|*} structuredError
+ * @returns {Promise.<{documentBody: *, errorMessages:string[], validation: boolean}>}
+ */
+Validation.prototype.validationPromise = function (index, collection, controller, action, id, documentCandidate, structuredError) {
+  var
+    // Avoid side effects on the requestObject during the validation
+    requestBodyClone = JSON.parse(JSON.stringify(documentCandidate)),
```

Comment:
and? is it a problem?
My concern is that the serialization process is costly and if it can be avoided, I'd rather to.